Api provides client with products

sealfury opened the pull request:
[API can provide client with collection of products](https://www.pivotaltracker.com/n/projects/2477606)
### User Story
```
As a restaurant API
In order to serve a client with products
I would like to be able to provide an endpoint for a products index
```
### Changes proposed in this pull request:
- Creating product model and spec test
- Creating products controller and routes, request spec test
- Changes from setup branch also included
### What we have learned working on this feature:
- To generate a products model/db/controller in rails and write tests for these

sealfury committed:
turns off generators for helpers, assets, and all specs except model specs

sealfury committed:
deleting demo request test, creating api_provides_client spec test file, generating prodcut model, migrating model db

sealfury committed:
creating product model spec test, dropping/creating/migrating db, adding validations to product model for name and price

sealfury committed:
creating products controller and route to products index, all tests green

sealfury committed:
fixing typo in request test

emtalen commented on app/controllers/api/products_controller.rb:
```diff
@@ -0,0 +1,7 @@
+class Api::ProductsController < ApplicationController
+  def index
+    product = Product.all 
```
please change the variable to plural, it is a more telling name since we are storing an array of several products in that variable

sealfury replied:
variable is now plural

emtalen commented on config/routes.rb:
```diff
@@ -1,7 +1,8 @@
-    namespace :v0 do
-      resources :pings, only: [:index], constraints: { format: 'json' }
-    end
+    get 'products/index'
```
remove this please, we don't declare routes like this

sealfury replied:
removed this route declaration and extra namespace that contained it

emtalen commented on config/routes.rb:
```diff
@@ -1,7 +1,8 @@
-    end
+    get 'products/index'
+  end
+  namespace :api do
```
You don't need two different namespaces for `api`

emtalen commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,20 @@
+
+RSpec.describe Product, type: :model do
+  describe '\'Products\' has db columns' do
+    it {is_expected.to have_db_column :id}
```
You don't need to test for the ID, every instance of a database table will get an ID because it is the primary key, so it is unnecessary to test for. It happens wether we want it or not.

sealfury replied:
removed test for id and added one for description

emtalen commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  describe 'Factory' do
+    it 'should have valid Factory' do
```
Keep the consistency with the declaration messages, like we touched upon yesterday in the demo, use `is expected to` instead of should

emtalen commented on spec/requests/api/v1/api_provides_client_with_collection_of_products_spec.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+
+    it 'should return two products' do
+      expect(JSON.parse(response.body)['products'].count).to eq 2
```
Please add the `response_json` module like you do in the Cooper challenge

sealfury replied:
added

emtalen replied:
Then please refactor this to `response_json` instead of `JSON.parse(response.body)` 🦖

sealfury replied:
refactored

sealfury committed:
adding validation for presence of description, test for validation
remove test for id db column
deleted automatically generated namespace/route
changing variable in product_controller to plural
correcting phrasing in spec declaration messages
adding response_json module

emtalen commented on spec/rails_helper.rb:
```diff
@@ -19,4 +19,5 @@
   config.filter_rails_from_backtrace!
   config.include FactoryBot::Syntax::Methods
   config.include(Shoulda::Matchers::ActiveRecord, type: :model)
+  config.include ResponseJSON
```
You need to create this module in your support folder

sealfury replied:
created

sealfury committed:
Refactoring request test to use response_json module
adding response_json module to support folder

sealfury committed:
changing url in line 5 of request spec test
